File: bao-source/bao-edge/src/contracts/operator-verification-report.ts
/**
 * Canonical app-owned verification report contract shared by mobile apps and control-plane readers.
 */

/** Current schema version for operator verification reports. */
export const OPERATOR_VERIFICATION_REPORT_SCHEMA_VERSION = "1.0" as const;

/** Platforms that emit operator verification reports. */
export type OperatorVerificationPlatform = "android" | "ios";

/** JSON-like input accepted at the report boundary before contract normalization. */
type JsonInput =
  | JsonScalar
  | { readonly [key: string]: JsonInput | undefined }
  | readonly JsonInput[];

/** Stable workflow scenarios covered by app-owned verification reports. */
export type OperatorVerificationScenario =
  | "cloud_chat"
  | "local_device_ai"
  | "calendar_native_action"
  | "web_research_rpa"
  | "automation_schedule";

/** Trigger source for one verification report run. */
export type OperatorVerificationTriggerSource =
  | "user"
  | "launch_injection"
  | "background_schedule"
  | "device_ai_protocol"
  | "native_intent";

/** Capability usage resolved for one verification run. */
export type OperatorVerificationUsage =
  | "chat"
  | "automation"
  | "research_rpa"
  | "native_action"
  | "speech_input"
  | "speech_output"
  | "local_device_ai";

/** Runtime transport used for one verification run. */
export type OperatorVerificationRuntimeTransport = "local" | "cloud";

/** Terminal state written into verification reports. */
export type OperatorVerificationTerminalState =
  | "idle"
  | "loading"
  | "success"
  | "empty"
  | "error-retryable"
  | "error-non-retryable"
  | "unauthorized";

/** Structured provenance for one runtime binding used during verification. */
export interface OperatorVerificationRuntimeBinding {
  /** Capability usage requested by the app for this run. */
  usage: OperatorVerificationUsage;
  /** Whether the app resolved to local or cloud execution. */
  transport: OperatorVerificationRuntimeTransport;
  /** Optional provider identifier for cloud-backed execution. */
  provider?: string;
  /** Optional model identifier for the resolved runtime. */
  model?: string;
  /** Optional model source or inventory family. */
  source?: string;
  /** Optional local model identifier when the run used a native model. */
  localModelName?: string;
  /** Optional target platform when the runtime dispatches remote agent work. */
  targetPlatform?: "android" | "ios";
}

/** Human-readable request summary emitted by the app. */
export interface OperatorVerificationRequestSummary {
  /** Prompt or high-level intent summary for the run. */
  prompt: string;
  /** Optional machine-readable metadata attached to the request. */
  metadata?: Record<string, string>;
}

/** Human-readable terminal reply summary emitted by the app. */
export interface OperatorVerificationReplySummary {
  /** Final user-visible reply or status summary. */
  message: string;
  /** Terminal state surfaced by the app. */
  state: OperatorVerificationTerminalState;
  /** Provenance of the reply content. */
  provenance: "cloud" | "local" | "native" | "xctest" | "rpa";
  /** Optional live source URLs used to produce the reply. */
  sources?: string[];
}

/** Structured evidence row proving one part of the verified workflow. */
export interface OperatorVerificationEvidence {
  /** Stable evidence category. */
  kind:
    | "url_visit"
    | "screenshot"
    | "ui_hierarchy"
    | "calendar_event"
    | "device_ai_report"
    | "flow_report"
    | "conversation_persisted";
  /** Outcome for this evidence row. */
  status: "success" | "failed" | "partial";
  /** Human-readable summary. */
  summary: string;
  /** Optional primary value associated with this evidence row. */
  value?: string;
  /** Optional absolute artifact path. */
  artifactPath?: string;
  /** Optional structured metadata emitted by the app. */
  metadata?: Record<string, string>;
}

/** Persisted entity identifiers created or touched by the run. */
export interface OperatorVerificationTargetIds {
  /** Optional application identifier under test. */
  appId?: string;
  /** Optional schedule identifier for durable automations. */
  scheduleId?: string;
  /** Optional calendar event identifier. */
  eventId?: string;
  /** Optional calendar identifier/title. */
  calendarId?: string;
  /** Optional conversation/thread identifier. */
  conversationId?: string;
  /** Optional emulator/simulator device identifier. */
  deviceId?: string;
}

/** Canonical app-owned operator verification report. */
export interface OperatorVerificationReport {
  /** Schema version for compatibility checks. */
  schemaVersion: typeof OPERATOR_VERIFICATION_REPORT_SCHEMA_VERSION;
  /** Correlation id for the verification run. */
  correlationId: string;
  /** Platform that emitted the report. */
  platform: OperatorVerificationPlatform;
  /** Stable scenario identifier. */
  scenario: OperatorVerificationScenario;
  /** Runtime binding used by the app for the run. */
  runtime: OperatorVerificationRuntimeBinding;
  /** Trigger source that started the run. */
  triggerSource: OperatorVerificationTriggerSource;
  /** Human-readable request summary. */
  request: OperatorVerificationRequestSummary;
  /** Terminal verification state. */
  terminalState: OperatorVerificationTerminalState;
  /** Human-readable reply summary. */
  reply: OperatorVerificationReplySummary;
  /** Ordered evidence rows captured by the app. */
  evidence: OperatorVerificationEvidence[];
  /** Persisted target identifiers created or touched by the app. */
  targetIds?: OperatorVerificationTargetIds;
  /** Absolute artifact paths produced by the run. */
  artifactPaths?: string[];
  /** ISO8601 start timestamp. */
  startedAt: string;
  /** ISO8601 completion timestamp. */
  completedAt: string;
}

// Parser helpers

type JsonScalar = string | number | boolean | null;
type JsonValue = JsonScalar | JsonRecord | readonly JsonValue[];
type JsonRecord = { readonly [key: string]: JsonValue };

function isJsonRecord(value: JsonInput | undefined): value is JsonRecord {
  return (
    value !== null && value !== undefined && typeof value === "object" && !Array.isArray(value)
  );
}

function asTrimmedString(value: JsonInput | undefined): string | null {
  return typeof value === "string" && value.trim().length > 0 ? value.trim() : null;
}

function asIsoTimestamp(value: JsonInput | undefined): string | null {
  const normalized = asTrimmedString(value);
  if (!normalized) {
    return null;
  }
  return Number.isNaN(Date.parse(normalized)) ? null : normalized;
}

function asMetadataRecord(value: JsonInput | undefined): Record<string, string> | undefined {
  if (!isJsonRecord(value)) {
    return undefined;
  }
  const entries = Object.entries(value)
    .map(([key, metadataValue]) => {
      const normalizedValue = asTrimmedString(metadataValue);
      return normalizedValue ? ([key, normalizedValue] as const) : null;
    })
    .filter((entry): entry is readonly [string, string] => entry !== null);
  return entries.length > 0 ? Object.fromEntries(entries) : undefined;
}

const VALID_PLATFORMS = new Set<string>(["android", "ios"]);
const VALID_SCENARIOS = new Set<string>([
  "cloud_chat",
  "local_device_ai",
  "calendar_native_action",
  "web_research_rpa",
  "automation_schedule",
]);
const VALID_TRIGGER_SOURCES = new Set<string>([
  "user",
  "launch_injection",
  "background_schedule",
  "device_ai_protocol",
  "native_intent",
]);
const VALID_TERMINAL_STATES = new Set<string>([
  "idle",
  "loading",
  "success",
  "empty",
  "error-retryable",
  "error-non-retryable",
  "unauthorized",
]);
const VALID_USAGES = new Set<string>([
  "chat",
  "automation",
  "research_rpa",
  "native_action",
  "speech_input",
  "speech_output",
  "local_device_ai",
]);
const VALID_TRANSPORTS = new Set<string>(["local", "cloud"]);
const VALID_EVIDENCE_KINDS = new Set<string>([
  "url_visit",
  "screenshot",
  "ui_hierarchy",
  "calendar_event",
  "device_ai_report",
  "flow_report",
  "conversation_persisted",
]);
const VALID_EVIDENCE_STATUSES = new Set<string>(["success", "failed", "partial"]);
const VALID_PROVENANCES = new Set<string>(["cloud", "local", "native", "xctest", "rpa"]);

function readOptionalTrimmedString(value: JsonInput | undefined): string | null {
  return asTrimmedString(value);
}

function readRequiredRecord(value: JsonInput | undefined): JsonRecord | null {
  return isJsonRecord(value) ? value : null;
}

function isValidMember<T extends string>(
  value: string | null,
  members: ReadonlySet<string>,
): value is T {
  return value !== null && members.has(value);
}

function buildOperatorVerificationRuntime(
  runtime: JsonRecord,
): OperatorVerificationRuntimeBinding | null {
  const usage = readOptionalTrimmedString(runtime.usage);
  const transport = readOptionalTrimmedString(runtime.transport);
  if (
    !(
      isValidMember<OperatorVerificationUsage>(usage, VALID_USAGES) &&
      isValidMember<OperatorVerificationRuntimeTransport>(transport, VALID_TRANSPORTS)
    )
  ) {
    return null;
  }

  const provider = readOptionalTrimmedString(runtime.provider);
  const model = readOptionalTrimmedString(runtime.model);
  const source = readOptionalTrimmedString(runtime.source);
  const localModelName = readOptionalTrimmedString(runtime.localModelName);
  const targetPlatform = readOptionalTrimmedString(runtime.targetPlatform);

  return {
    usage,
    transport,
    ...(provider ? { provider } : {}),
    ...(model ? { model } : {}),
    ...(source ? { source } : {}),
    ...(localModelName ? { localModelName } : {}),
    ...(targetPlatform && VALID_PLATFORMS.has(targetPlatform)
      ? { targetPlatform: targetPlatform as "android" | "ios" }
      : {}),
  };
}

function buildOperatorVerificationRequest(
  request: JsonRecord,
): OperatorVerificationRequestSummary | null {
  const prompt = readOptionalTrimmedString(request.prompt);
  if (!prompt) {
    return null;
  }
  const metadata = asMetadataRecord(request.metadata);
  return {
    prompt,
    ...(metadata ? { metadata } : {}),
  };
}

function buildOperatorVerificationReply(
  reply: JsonRecord,
): OperatorVerificationReplySummary | null {
  const message = readOptionalTrimmedString(reply.message);
  const state = readOptionalTrimmedString(reply.state);
  const provenance = readOptionalTrimmedString(reply.provenance);
  if (
    !(
      message &&
      state &&
      VALID_TERMINAL_STATES.has(state) &&
      provenance &&
      VALID_PROVENANCES.has(provenance)
    )
  ) {
    return null;
  }

  return {
    message,
    state: state as OperatorVerificationTerminalState,
    provenance: provenance as OperatorVerificationReplySummary["provenance"],
    ...(Array.isArray(reply.sources)
      ? {
          sources: reply.sources.filter(
            (v): v is string => typeof v === "string" && v.trim().length > 0,
          ),
        }
      : {}),
  };
}

function buildOperatorVerificationTargetIds(targetIds: JsonRecord): OperatorVerificationTargetIds {
  const appId = readOptionalTrimmedString(targetIds.appId);
  const scheduleId = readOptionalTrimmedString(targetIds.scheduleId);
  const eventId = readOptionalTrimmedString(targetIds.eventId);
  const calendarId = readOptionalTrimmedString(targetIds.calendarId);
  const conversationId = readOptionalTrimmedString(targetIds.conversationId);
  const deviceId = readOptionalTrimmedString(targetIds.deviceId);

  return {
    ...(appId ? { appId } : {}),
    ...(scheduleId ? { scheduleId } : {}),
    ...(eventId ? { eventId } : {}),
    ...(calendarId ? { calendarId } : {}),
    ...(conversationId ? { conversationId } : {}),
    ...(deviceId ? { deviceId } : {}),
  };
}

function buildOperatorVerificationArtifactPaths(value: JsonInput | undefined): string[] {
  return Array.isArray(value)
    ? value.filter((entry): entry is string => typeof entry === "string" && entry.trim().length > 0)
    : [];
}

function buildOperatorVerificationEvidenceRows(
  value: JsonInput | undefined,
): OperatorVerificationEvidence[] {
  return Array.isArray(value)
    ? value
        .map((entry): OperatorVerificationEvidence | null => parseEvidenceRow(entry))
        .filter((row): row is OperatorVerificationEvidence => row !== null)
    : [];
}

function parseEvidenceRow(value: JsonInput | undefined): OperatorVerificationEvidence | null {
  if (!isJsonRecord(value)) {
    return null;
  }
  const kind = asTrimmedString(value.kind);
  const status = asTrimmedString(value.status);
  const summary = asTrimmedString(value.summary);
  const detailValue = asTrimmedString(value.value);
  const artifactPath = asTrimmedString(value.artifactPath);
  const metadata = asMetadataRecord(value.metadata);
  if (
    !(
      kind &&
      status &&
      summary &&
      VALID_EVIDENCE_KINDS.has(kind) &&
      VALID_EVIDENCE_STATUSES.has(status)
    )
  ) {
    return null;
  }
  return {
    kind: kind as OperatorVerificationEvidence["kind"],
    status: status as OperatorVerificationEvidence["status"],
    summary,
    ...(detailValue ? { value: detailValue } : {}),
    ...(artifactPath ? { artifactPath } : {}),
    ...(metadata ? { metadata } : {}),
  };
}

/** Parse and validate a JSON-like payload as a canonical operator verification report. */
export function parseOperatorVerificationReport(
  value: JsonInput | undefined,
): OperatorVerificationReport | null {
  const record = readRequiredRecord(value);
  if (!record) {
    return null;
  }

  const schemaVersion = readOptionalTrimmedString(record.schemaVersion);
  const correlationId = readOptionalTrimmedString(record.correlationId);
  const platform = readOptionalTrimmedString(record.platform);
  const scenario = readOptionalTrimmedString(record.scenario);
  const triggerSource = readOptionalTrimmedString(record.triggerSource);
  const terminalState = readOptionalTrimmedString(record.terminalState);
  const startedAt = asIsoTimestamp(record.startedAt);
  const completedAt = asIsoTimestamp(record.completedAt);
  const runtime = readRequiredRecord(record.runtime);
  const request = readRequiredRecord(record.request);
  const reply = readRequiredRecord(record.reply);

  if (
    schemaVersion !== OPERATOR_VERIFICATION_REPORT_SCHEMA_VERSION ||
    !correlationId ||
    !isValidMember<OperatorVerificationPlatform>(platform, VALID_PLATFORMS) ||
    !isValidMember<OperatorVerificationScenario>(scenario, VALID_SCENARIOS) ||
    !isValidMember<OperatorVerificationTriggerSource>(triggerSource, VALID_TRIGGER_SOURCES) ||
    !isValidMember<OperatorVerificationTerminalState>(terminalState, VALID_TERMINAL_STATES) ||
    !startedAt ||
    !completedAt ||
    !runtime ||
    !request ||
    !reply
  ) {
    return null;
  }

  const parsedRuntime = buildOperatorVerificationRuntime(runtime);
  const parsedRequest = buildOperatorVerificationRequest(request);
  const parsedReply = buildOperatorVerificationReply(reply);
  if (!(parsedRuntime && parsedRequest && parsedReply)) {
    return null;
  }

  const targetIds = readRequiredRecord(record.targetIds);
  const artifactPaths = buildOperatorVerificationArtifactPaths(record.artifactPaths);
  const evidence = buildOperatorVerificationEvidenceRows(record.evidence);

  return {
    schemaVersion: OPERATOR_VERIFICATION_REPORT_SCHEMA_VERSION,
    correlationId,
    platform,
    scenario,
    runtime: parsedRuntime,
    triggerSource,
    request: parsedRequest,
    terminalState,
    reply: parsedReply,
    evidence,
    ...(targetIds ? { targetIds: buildOperatorVerificationTargetIds(targetIds) } : {}),
    ...(artifactPaths.length > 0 ? { artifactPaths } : {}),
    startedAt,
    completedAt,
  };
}

File: bao-source/bao-edge/iOS/BaoEdge/Sources/BaoEdgeCore/OperatorVerificationReport.swift
import Foundation

/// Current schema version for app-owned operator verification reports.
public let operatorVerificationReportSchemaVersion = "1.0"

/// Stable workflow scenarios emitted by mobile verification reports.
public enum OperatorVerificationScenario: String, Codable, Sendable {
    case cloudChat = "cloud_chat"
    case localDeviceAi = "local_device_ai"
    case calendarNativeAction = "calendar_native_action"
    case webResearchRpa = "web_research_rpa"
    case automationSchedule = "automation_schedule"
}

/// Stable capability usages emitted by mobile verification reports.
public enum OperatorVerificationUsage: String, Codable, Sendable {
    case chat
    case automation
    case researchRpa = "research_rpa"
    case nativeAction = "native_action"
    case speechInput = "speech_input"
    case speechOutput = "speech_output"
    case localDeviceAi = "local_device_ai"
}

/// Trigger sources for app-owned verification runs.
public enum OperatorVerificationTriggerSource: String, Codable, Sendable {
    case user
    case launchInjection = "launch_injection"
    case backgroundSchedule = "background_schedule"
    case deviceAiProtocol = "device_ai_protocol"
    case nativeIntent = "native_intent"
}

/// Runtime transport used during one verification run.
public enum OperatorVerificationRuntimeTransport: String, Codable, Sendable {
    case local
    case cloud
}

/// Terminal states shared by app-owned verification reports.
public enum OperatorVerificationTerminalState: String, Codable, Sendable {
    case idle
    case loading
    case success
    case empty
    case errorRetryable = "error-retryable"
    case errorNonRetryable = "error-non-retryable"
    case unauthorized

    /// Maps a flow execution state into the canonical verification terminal state.
    public init(flowState: FlowExecutionState) {
        switch flowState {
        case .idle: self = .idle
        case .loading: self = .loading
        case .success: self = .success
        case .empty: self = .empty
        case .errorRetryable: self = .errorRetryable
        case .errorNonRetryable: self = .errorNonRetryable
        case .unauthorized: self = .unauthorized
        }
    }
}

/// Runtime binding metadata written into app-owned verification reports.
public struct OperatorVerificationRuntimeBinding: Codable, Sendable {
    public var usage: OperatorVerificationUsage
    public var transport: OperatorVerificationRuntimeTransport
    public var provider: String?
    public var model: String?
    public var source: String?
    public var localModelName: String?
    public var targetPlatform: String?

    public init(
        usage: OperatorVerificationUsage,
        transport: OperatorVerificationRuntimeTransport,
        provider: String? = nil,
        model: String? = nil,
        source: String? = nil,
        localModelName: String? = nil,
        targetPlatform: String? = nil
    ) {
        self.usage = usage
        self.transport = transport
        self.provider = provider
        self.model = model
        self.source = source
        self.localModelName = localModelName
        self.targetPlatform = targetPlatform
    }
}

/// Human-readable request summary written into app-owned verification reports.
public struct OperatorVerificationRequestSummary: Codable, Sendable {
    public var prompt: String
    public var metadata: [String: String]

    public init(prompt: String, metadata: [String: String] = [:]) {
        self.prompt = prompt
        self.metadata = metadata
    }
}

/// Human-readable reply summary written into app-owned verification reports.
public struct OperatorVerificationReplySummary: Codable, Sendable {
    public var message: String
    public var state: OperatorVerificationTerminalState
    public var provenance: String
    public var sources: [String]

    public init(
        message: String,
        state: OperatorVerificationTerminalState,
        provenance: String,
        sources: [String] = []
    ) {
        self.message = message
        self.state = state
        self.provenance = provenance
        self.sources = sources
    }
}

/// Structured evidence row proving one part of a verified workflow.
public struct OperatorVerificationEvidence: Codable, Sendable {
    public var kind: String
    public var status: String
    public var summary: String
    public var value: String?
    public var artifactPath: String?
    public var metadata: [String: String]

    public init(
        kind: String,
        status: String,
        summary: String,
        value: String? = nil,
        artifactPath: String? = nil,
        metadata: [String: String] = [:]
    ) {
        self.kind = kind
        self.status = status
        self.summary = summary
        self.value = value
        self.artifactPath = artifactPath
        self.metadata = metadata
    }
}

/// Persisted target identifiers created or touched by the run.
public struct OperatorVerificationTargetIds: Codable, Sendable {
    public var appId: String?
    public var scheduleId: String?
    public var eventId: String?
    public var calendarId: String?
    public var conversationId: String?
    public var deviceId: String?

    public init(
        appId: String? = nil,
        scheduleId: String? = nil,
        eventId: String? = nil,
        calendarId: String? = nil,
        conversationId: String? = nil,
        deviceId: String? = nil
    ) {
        self.appId = appId
        self.scheduleId = scheduleId
        self.eventId = eventId
        self.calendarId = calendarId
        self.conversationId = conversationId
        self.deviceId = deviceId
    }
}

/// Canonical app-owned verification report.
public struct OperatorVerificationReport: Codable, Sendable {
    public var schemaVersion: String
    public var correlationId: String
    public var platform: String
    public var scenario: OperatorVerificationScenario
    public var runtime: OperatorVerificationRuntimeBinding
    public var triggerSource: OperatorVerificationTriggerSource
    public var request: OperatorVerificationRequestSummary
    public var terminalState: OperatorVerificationTerminalState
    public var reply: OperatorVerificationReplySummary
    public var evidence: [OperatorVerificationEvidence]
    public var targetIds: OperatorVerificationTargetIds?
    public var artifactPaths: [String]
    public var startedAt: String
    public var completedAt: String

    public init(
        correlationId: String,
        platform: String = "ios",
        scenario: OperatorVerificationScenario,
        runtime: OperatorVerificationRuntimeBinding,
        triggerSource: OperatorVerificationTriggerSource,
        request: OperatorVerificationRequestSummary,
        terminalState: OperatorVerificationTerminalState,
        reply: OperatorVerificationReplySummary,
        evidence: [OperatorVerificationEvidence] = [],
        targetIds: OperatorVerificationTargetIds? = nil,
        artifactPaths: [String] = [],
        startedAt: String,
        completedAt: String
    ) {
        self.schemaVersion = operatorVerificationReportSchemaVersion
        self.correlationId = correlationId
        self.platform = platform
        self.scenario = scenario
        self.runtime = runtime
        self.triggerSource = triggerSource
        self.request = request
        self.terminalState = terminalState
        self.reply = reply
        self.evidence = evidence
        self.targetIds = targetIds
        self.artifactPaths = artifactPaths
        self.startedAt = startedAt
        self.completedAt = completedAt
    }
}

/// File-backed persistence contract for app-owned verification reports.
public protocol OperatorVerificationReporting: Sendable {
    /// Persist the latest verification report and an archived per-run copy.
    @discardableResult
    func persist(_ report: OperatorVerificationReport) throws -> URL
}

/// Default file-backed verification reporter used by iOS operator flows.
public struct FileOperatorVerificationReporter: OperatorVerificationReporting, @unchecked Sendable {
    private let fileManager: FileManager
    private let rootDirectoryURL: URL
    private let encoder = JSONEncoder()

    public init(
        fileManager: FileManager = .default,
        rootDirectoryURL: URL? = nil
    ) {
        self.fileManager = fileManager
        let baseDirectory = rootDirectoryURL
            ?? fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
                .appendingPathComponent("bao-edge-operator-verification", isDirectory: true)
        self.rootDirectoryURL = baseDirectory
    }

    @discardableResult
    public func persist(_ report: OperatorVerificationReport) throws -> URL {
        let runsDirectory = rootDirectoryURL.appendingPathComponent("runs", isDirectory: true)
        try ensureDirectory(rootDirectoryURL)
        try ensureDirectory(runsDirectory)
        let archivedURL = runsDirectory.appendingPathComponent("\(report.correlationId).json", isDirectory: false)
        let latestURL = rootDirectoryURL.appendingPathComponent("latest.json", isDirectory: false)
        let data = try encoder.encode(report)
        try data.write(to: archivedURL, options: [.atomic])
        try data.write(to: latestURL, options: [.atomic])
        return latestURL
    }

    private func ensureDirectory(_ url: URL) throws {
        if fileManager.fileExists(atPath: url.path) {
            return
        }
        try fileManager.createDirectory(at: url, withIntermediateDirectories: true)
    }
}
